### ISSUE TITLE
Script doesn't work in Firemonkey 

### ISSUE THREAD

**Issue by rubaboo** [erosman/support#277]
Could you have a look at this script to see if it can be made to work with Firemonkey with some adjustments? It works in Tampermonkey…
https://raw.githubusercontent.com/Wizmann/Utils/master/UserScript/WorkflowyPlus.js

**Comment by erosman** [comment]
Script uses JQuery functions without `@require` any JQuery.

I don't know which JQuery version it uses but try with JQeury 3 and see

```
// ==UserScript==
// @name         WorkflowyPlus
// @namespace    http://wizmann.tk
// @version      0.1
// @author       Wizmann
// @match        https://workflowy.com/*
// @grant        none
// @require      jquery-3
// ==/UserScript==
```

**Comment by rubaboo** [comment]
Thanks, the script works with `@require`.

This is probably not very helpful, but if you want details:
I did not check it with other M's. There are no instructions at https://github.com/Wizmann/Utils, but https://blog.workflowy.com/2016/01/06/inline-images/ suggests using TM for the script (without explaining why TM and not any other M).

**Comment by rubaboo** [comment]
Actually, I do have one more question. I have uBlock Origin installed, and this script still wouldn't work until I allowed Google Analytics: `workflowy.com google-analytics.com * allow`. Any idea why that is?

**Comment by erosman** [comment]
> TM for the script (without explaining why TM and not any other M).

> [Tampermonkey • Documentation](https://www.tampermonkey.net/documentation.php)
> 
> If @grant is followed by 'none' the sandbox is disabled and the script will run directly at the page context. In this mode no GM_* function but the GM_info property will be available.

As I suspected, the script in TM runs in page context.

**Comment by erosman** [comment]
> [Tampermonkey • Documentation](https://www.tampermonkey.net/documentation.php)
> 
> If @grant is followed by 'none' the sandbox is disabled and the script will run directly at the page context. In this mode no GM_* function but the GM_info property will be available.

As I suspected, the script in TM runs in page context (which has security concerns).

If you want to mimic the TM and inject the script into page context, then this is what you can do ...

📌  `GM.addScript()` is _FireMonkey_ only.


```js
// ==UserScript==
// @name         WorkflowyPlus
// @namespace    http://wizmann.tk
// @version      0.1
// @author       Wizmann
// @match        https://workflowy.com/*
// @grant        none
// ==/UserScript==
/* jshint -W097 */
'use strict';

GM.addScript(`

function do_parseImg() {
    console.log("do_parseImg");
    $(this).nextAll(".content-img").remove();
    var lines = $(this).text().split("\n");
    var img_re = /^\!\[(?<property>.*?)(?:,(?<ratio>\d+))?\]\((?<url>.+)\)$/;

    for (var i = 0; i < lines.length; i++) {
        var line = lines[i].trim();
        var img = line.match(img_re);
        if (img === null) {
            continue;
        }
        var property = img.groups.property;
        var img_url = img.groups.url

        // console.log(property, img_url);

        if (property === "iframe") {
            $(this).after('<div class="content-img"><iframe width="100%" height="512" src="' + img_url + '" frameborder="0" allowfullscreen=""></iframe></div>');
        } else if (property === "audio") {
            $(this).after('<div class="content-img"><audio src="' + img_url + '" controls="controls"></audio></div>');
        } else {
            console.log(img);
            if (img.groups.ratio != undefined) {
                $(this).after('<div class="content-img"><img width="' + img.groups.ratio + '%" ' + 'src="' + img_url + '"/></div>');
            } else {
                $(this).after('<div class="content-img"><img src="' + img_url + '"/></div>');
            }
        }
    }
}

function parseImg() {
    console.log("parseImg");
    // console.log($("div.notes div.content"));
    // $("div.notes div.content").keyup(do_parseImg);
    // $("div.notes div.content").click(do_parseImg);
    $("div.notes div.content").each(do_parseImg);
};

$(window).bind("load hashchange", parseImg);
window.addEventListener('popstate', parseImg);

window.WFEventListener = event => {
  switch (event) {
    case 'documentReady':
    case 'expand':    // via click on + & Ctrl+Space on child only. NOT SUPPORTED: Ctrl+Down, Ctrl+Space on parent & menu
    case 'collapse':    // via click on - & Ctrl+Space on child only. NOT SUPPORTED: Ctrl+Up, Ctrl+Space on parent & menu
          parseImg(); break;
    case 'operation--move':    // moving any bullet, via mouse or KBS, including indent, outdent
    case 'operation--edit':    // automatic save after any editing text operations. Usually fires within 1s of last keystroke
          console.log(event);
          var node = WF.focusedItem().getElement();
          console.log($(node).find("div.notes div.content"));
          $(node).find("div.notes div.content").each(do_parseImg);
          break;
  }
};

var pushState = history.pushState;
history.pushState = function () {
    pushState.apply(history, arguments);
    parseImg();
};

// end of GM.addScript
`);
```

**Comment by erosman** [comment]
I am implementing a new feature to cater for this TM behaviour.

When testing, the above script still runs into error since `https://workflowy.com/` does have JQuery. Does is function fully in TM?

**Comment by rubaboo** [comment]
Yes, in TM it just works. I tried in a different profile though, one that doesn't have uBO. I can try with uBO next. Did you notice my question about Google Analytics? Took me a while to find what to unblock, GA was the last thing on my list to try unblocking.

**Comment by rubaboo** [comment]
@erosman I think I'm onto something. Tried TM+uBO, the script works, no need to unblock Google Analytics. That means FM is somehow requiring GA unblocked in order to work.

**Comment by erosman** [comment]
I get the same error in TM as in FM and that is JavaScript parsing error (Developer Tools `Ctrl +Shift + j`)

![image](https://user-images.githubusercontent.com/6458876/105092681-ccb69e80-5ab6-11eb-9648-2f1e216b2e0a.png)

That is the issue with the script running. `$` & `$(this).nextAll` etc are all JQuery function. If there is no JQuery in that page, there will be an error.
I tested it on `https://workflowy.com/`

I cant test it in inner page as it requires login. It is possible that inner pages have JQuery.
Is there a page that I can test on that doesn't require login?

Google Analytics shouldn't have any bearing on it as the script doesn't use GA.

**Comment by rubaboo** [comment]
I think we are discussing 2, if not 3 issues here. I can create separate issues for clarity if needed.
1. The initial issue - script not working in FM. Fixed by adding `// @require      jquery-3` to the script. FIXED.
2. You would like to add a new feature, to improve compatibility with TM scripts(?). This one is over my head. No comment.
3. The FIXED (see #1) script does not produce a desired result unless I unblock GA in uBO. With TM, no unblocking required. I know GA should not have any bearing, but the issue is there and is repeatable. **This is the issue I am currently concerned about.** I will see about sharing an inner page with you where I can demo the issue. If I need to send you a secret URL, how can I do that?

**Comment by erosman** [comment]
### Context
TM injects the script with no `@grant `into **page context** while FM in the **userscirpt context** & GM|VM into **content context**.

The behaviour of the script will be different in different context.
The new feature I am working will mimic TM if a new flag is added... that is for future.


### JQuery
Above script requires JQuery to run but it doesn't have it.
If the script works with TM without JQuery, that would mean the page that it is injected to by TM already has JQeury which the script can use.

In case of current FM, script can be injected in the page, same as TM by using the code given in https://github.com/erosman/support/issues/277#issuecomment-762837164

That way, there should be no need to add `// @require      jquery-3`

### uBlock Origin
I am guessing there are issues that I cant see form the front page that I tried.

### Test URL
Do you visit `https://chat.mozilla.org/`?
If so, you can send me a private message.
You can find my name (even when I am not online) in `https://chat.mozilla.org/#/room/#addon-reviewers:mozilla.org`

**Comment by rubaboo** [comment]
Sent a DM.

**Comment by erosman** [comment]
I checked the sample page.

- `$` & `$(this).nextAll` are defined in` document_view.min.js` which seems to have the code from JQeury 3.5.0 embedded in it.
- TM injects into page context and then the userscript use above file to run its functions.
- If you use the https://github.com/erosman/support/issues/277#issuecomment-762837164 version without JQuery, the script should work fine. I am adding a new feature to FireMonkey to automate this with a metadata flag.
- I have uBO blocking GA and it works

**Comment by rubaboo** [comment]
IDK, the injection way does not work for me at all. Since you don't have the same problem, I guess it's up to me to find out why I'm having this issue. Just one more thing - when you were testing, did you replace your uBO settings with the settings I uploaded to the chat? They are just "medium-mode" settings with some additional filters.

**Comment by erosman** [comment]
> when you were testing, did you replace your uBO settings with the settings I uploaded to the chat? They are just "medium-mode" settings with some additional filters.

No... My uBO is almost the default setting. Try the default and see.

**Comment by rubaboo** [comment]
I give up. Tried everything. I don't have skills to find out where the problem is. Most puzzling is - in a brand new profile, with nothing but FM installed, the script does not work at all. Not with `// @require jquery-3`, nor injected. I'll leave the ticket to you to close or keep for your further research notes. (That's what I think some of your replies are - either notes to yourself, or you are greatly overestimating my proficiency with web programming.)
Thanks for your help.

Oh, I tried the script with all 4 Monkeys. GM - "as is" logs an error, with `@require` no errors but not working. TM, VM - works as is, no `@require` or injection needed.

**Comment by erosman** [comment]
> Oh, I tried the script with all 4 Monkeys. GM - "as is" logs an error, with @require no errors but not working. TM, VM - works as is, no @require or injection needed.

FM & GM inject into sandbox.
FM injects into dedicated userScript sandbox.
GM injects into content sandbox.

TM & VM inject into page.
TM injects into `page` if `@grant` is `none`.
VM injects into `page` if default injection mode is set to `auto` or `page` or `@inject-into` is set to `page`.
![image](https://user-images.githubusercontent.com/6458876/105328422-43ea5080-5be5-11eb-9c49-a29433bffdee.png)


The new release of FM will have a `@inject-into` option so you can set it to behave like TM & VM for this script. This feature is done.

I am in the middle of re-styling the options page to prepare for Android (it is a mess now). As soon as it is done, I will release the next version and you can test it.

I will keep this open as this is not the only script with `page` vs `sandbox` issue which results in a script working in TM|VM and not in FM|GM.

**Comment by erosman** [comment]
v2.13 with `@inject-into` support uploaded.

Try the script with it and let me know.

```
// ==UserScript==
// @name         WorkflowyPlus
// @namespace    http://wizmann.tk
// @version      0.1
// @author       Wizmann
// @match        https://workflowy.com/*
// @grant        none
// @inject-into  page
// ==/UserScript==
```

**Comment by rubaboo** [comment]
Yes, it works now. Thank you.

I don't feel this deserves an issue, and you might have this planned already - about the new layout of options page - it would be great if the sidebar was pinnable. In a desktop browser it makes sense to have it visible at all times, as before.

**Comment by pintassilgo** [comment]
`@inject-into` is a great improvement for compatibility, thank you. But is still required that we add it to scripts that need it. For scripts in active development, if the dev doesn't agree to add it, we're basically restricting FireMonkey to JS-skilled users, besides the hassle of having to fix scripts every time they update.

Violentmonkey uses "auto" injection mode by default, trying to detect if each script should be injected to content or to the page. It's efficient, I have both content and page scripts and all work fine even though none of them use `@inject-into`. Would it be possible to try to do something like that?

**Comment by erosman** [comment]
> Yes, it works now. Thank you.

Great... :+1: 

> I don't feel this deserves an issue, and you might have this planned already - about the new layout of options page - it would be great if the sidebar was pinnable. In a desktop browser it makes sense to have it visible at all times, as before.

There are pro & cons with both set-ups.
I had the sidebar panel at first since I liked to be able to quickly move through scripts.
On the other hand, on smaller or 4:3 screens, the wider area for the code is also beneficial.

Let's see.... :thinking: 

> `@inject-into` is a great improvement for compatibility, thank you. But is still required that we add it to scripts that need it. For scripts in active development, if the dev doesn't agree to add it, we're basically restricting FireMonkey to JS-skilled users, besides the hassle of having to fix scripts every time they update.

> What about adding a similar option? To be able to change the inject to content by default behavior.

@pintassilgo `@inject-into` is actually a metadata entry supported by ViolentMonkey (not supported by GM or TM).

- GM has no support for `page` mode.
- TM injects into `page` when `@grant none` without any GM function but with `GM_info`. Providing `GM_info` in `page` context means web sites will also have access to it which would allow them to track & block users and/or the script.
- VM injects into `page` **WITH** GM functions which is a considerable security risk as web sites will also have access to those GM functions

Furthermore, as mentioned, CSP of the page can block script injection. The CSP has to be checked manually via headers. VM injects manually (read https://github.com/erosman/support/issues/274#issuecomment-765878053) which is not done in FireMonkey.

VM has 3 injection options which are page, content, auto.
- `page` has above considerations
- `content` also has security consideration as it is privileged
- `auto` is `page` or `content` based on CSP

> what about adding a way to set injection mode per-script? Outside of userscript code, with higher priority than @inject-into. So fixed scripts wouldn't break again when updated by devs that don't care about FM. Just like any other userscript manager allow users to set custom @include/@exclude without touching the code.

That is possible but then bloats the preferences & code with a lot of new data.
Adding `@inject-into`  which is supported by FM|VM and ignored by GM|TM does not cause any hassle for the userscript developer. Doesn't that require just a tiny bit of effort?

Let's see :thinking: 

ATM, each script manger has features that are not compatible with others. Userscript developers would need to take that into account. FireMonkey user-base is still growing and there might be a time that userscript developers take notice.

**Comment by pintassilgo** [comment]
Thanks for detailed reply.

> * GM has no support for page mode.

Right. That's why I no longer use GM since v4.0. Incompatible with many scripts not written for it. I could fix scripts by myself, but when they update I'll need to fix them again and again... It doesn't sound right. I think of UserScripts as something almost universal, "cross-plataform". My personal view is that UJS managers should try hard to preserve compatibility.

> * VM injects into `page` **WITH** GM functions which is a considerable security risk as web sites will also have access to those GM functions

Only if the UserScript exposes those functions to the page, right? Like with `unsafeWindow.GM_getValue = GM_getValue` or `console.log(GM_getValue)`. Otherwise, even @grant'ing GM functions I can't access them with web console. If my reasoning is correct, the issue is in malicious/bad written UserScript. UserScripts have access to GM functions anyway, so I don't see much difference.

> Adding `@inject-into` which is supported by FM|VM and ignored by GM|TM does not cause any hassle for the userscript developer. Doesn't that require just a tiny bit of effort?

I agree with you, but we both know that we can't expect that everyone would be willing to do this tiny effort. Violentmonkey has less than a tenth of users compared to Tampermonkey. Effort and costs may be a little higher, but I think UJS managers should try to cover those things, mainly those who lack higher userbase (that's why Firefox need to implement Chromium web *standards* and not the other way around, as much as we don't like it).

But I perfectly understand that you think otherwise, fair enough, I'm just saying.

**Comment by erosman** [comment]
> My personal view is that UJS managers should try hard to preserve compatibility.

Should userscript mangers adapt to various ways userscript developers like to write the code that may only work in one manger and not in others? :thinking: 

In web design, it is the job of the web designer to develop content that works the same way in all browsers, not the other way round. Developing browser specific design has always been a sign of poor quality.  :wink: 

AFA compatibility, GM was the first and the pioneer. VM and TM did not preserve the compatibility in many aspect. TM & VM are not 100% compatible with each other either.
Result is that some scripts work in one and not another.

> Only if the UserScript exposes those functions to the page, right?

No... whatever is in `page` context is accessible by the page (web site). The whole point of different context is the separation of `page` & `content` & `browser`.

> Like with unsafeWindow.GM_getValue = GM_getValue or console.log(GM_getValue). Otherwise, even @grant'ing GM functions I can't access them with web console. 

**web console** has a different context which is `content` context.
Furthermore, each `content` context is isolated. `content` JavaScript injected by one add-on cant access `content` JavaScript injected by another add-on. There can be many `content` context  (i.e. sandboxes) in one page.

However, there is ONE `page` context which belongs to the **web site**. Whatever injected into that context, will also belong to the web site. That is the reason `page` XHR/fetch works fine in sending that site's credentials from `page` context but requires more parameters in GM functions.

From FireMonkey help ...

> **Extension JavaScript Context/Scope**
> 
> - Browser Context: Trusted privileged code to interact with browser
> - Content Context: Trusted extension JS injected into web page with API privileges
> - User-Script Context: Untrusted unverified 3rd party JS injected into web page without direct API privileges
> - Page Context: JS injected in a web page by the website


> UserScripts have access to GM functions anyway, so I don't see much difference.

Userscripts have access to GM functions in `userscript` context & `content` context BUT websites i.e. `page` context, don't have access to them.

**Note:** `content` context is not as secure as `userscript` context. `userscript` context has only access to the GM functions set by the script manger. `content` context, in addition to GM functions set by the script manager, also has access to a number of browser APIs (read https://github.com/erosman/support/issues/276#issuecomment-761763943). Usersript managers can implement measures to combat this but there are possible to be bypassed.

Communicating with `page` context is done through a method that is also accessible to web site. Web site can then hijack this channel. It is a two way bridge.

> But I perfectly understand that you think otherwise, fair enough, I'm just saying.

Feedback is good. I consider all feedbacks when developing. There are features now in FireMonkey that I personally wouldn't have included (and didn't at first). They were added after user feedback. :+1:

**Comment by pintassilgo** [comment]
I just noticed that FM's `@inject-into  page` is different from VM.

VM doesn't append the script directly in the page. FM exposes everything. With FM, if my UserScript has `let exposed = true` and doesn't use IIFE, `exposed` will be accessible on the page . VM do it in a way that I can't find the script in source, while FM simply append a `<script>`.

**Comment by erosman** [comment]
> VM do it in a way that I can't find the script in source, while FM simply append a <script>.

That is intentional. 

> VM doesn't append the script directly in the page.

Any method of injection does not make any difference. The injected will still be part of `page` JavaScript & accessible by the website.

> VM do it in a way that I can't find the script in source, while FM simply append a <script>.

It is also easy to hide it from the inspector BUT not from the `page` JavaScript.

Traditionally, they inject the JavaScript into `page` context and then remove the injected node.
After injection, JavaScript is parsed by the `page` and it will reside in the `page` JavaScript. Removing the node is only superficial.

I can remove the node like others but it makes absolutely no difference to what we discussed earlier.  The code is still accessible by the website.
I kept the node in the DOM so that user can visually inspect it.

**Comment by pintassilgo** [comment]
> In web design, it is the job of the web designer to develop content that works the same way in all browsers, not the other way round. Developing browser specific design has always been a sign of poor quality. 😉

True. But even so, everyday we see browsers (Firefox) needing to implement or fake non-standards things (from Blink) to not break sites to not lose users. Because at the end of the day that's how the world works, unfortunately.

As I said, I can fix scripts that are incompatible with FM. But I'm thinking on those who can't. And even those who can fix would need to reapply the patch every time the script updates. That's why at least I suggested an option to set @inject-into externally to the code, just like user-matches.

> GM was the first and the pioneer. VM and TM did not preserve the compatibility in many aspect. TM & VM are not 100% compatible with each other either.

You're not talking about GM 4, are you? I don't remember seeing any incompatibility with UserScripts written for GM 3-. The *de facto* standard was Greasemonkey 3-, not GM4. It was GM that jumped off the ship of preserving compatibility. And when it did that, GM was no longer the king, TM was already much more popular. So I think this example doesn't fit.

> Communicating with page context is done through a method that is also accessible to web site. Web site can then hijack this channel. It is a two way bridge.

How could a page run `GM.getValue`? Suppose an userScript that simply runs `GM.setValue('abc', 1)` (and defines @grant  `GM.getValue`, if needed to *leak* the function).

**Comment by pintassilgo** [comment]
> Any method of injection does not make any difference.

As I said, apparently there's a difference: VM's method doesn't expose userScript variables to the page, unlike FM's method. Maybe FM should wrap the code in a IIFE, that should de enough.

> I decided to remove the node to avoid confusion.

:wink:

**Comment by erosman** [comment]
> I don't remember seeing any incompatibility with UserScripts written for GM 3-

TM|VM had/have metadata entries that were/are not present in GM3 or GM4. There are other incompatibilities as well.
Read FireMonkey Help for more info.

> How could a page run GM.getValue? Suppose an userScript that simply runs GM.setValue('abc', 1) (and defines @grant GM.getValue, if needed to leak the function).

Read FireMonkey Help under **Xray Vision & Sharing objects with page scripts**. The `window` method used by userscript injected into page context is a global window method and webpage script also has access to it.

It is better not get too deep into the discussion of other userscript managers as it may be counterproductive.

Main objective of FireMonkey has always been security & performance.
Compromising the above 2 principles with new features, goes against the philosophy of the FireMonkey.  :wink:

**Comment by pintassilgo** [comment]
> TM|VM had/have metadata entries that were/are not present in GM3 or GM4.

But the few metadata entries supported by GM are also supported by every UJS manager. That's what I'm talking about, not the other way around. Every manager author thinks their extension is the best, it's OK to add extra features. But they all tried to keep compatibility with *standard/universal* metadata i.e. GM3.

Btw, I think VM table is wrong. For instance, apparently it supports [@updateURL](https://github.com/violentmonkey/violentmonkey/search?q=updateURL) and @author. The default metadata block when you click to create a new userScript even uses @author.

> The `window` method used by userscript injected into page context is a global window method and webpage script also has access to it.

Right, but I still don't know how a page could access GM* functions if they aren't explicitly exposed by the UserScript (like by `unsafeWindow.GM_getValue = GM_getValue` as I said).

> It is better not get too deep into the discussion of other userscript managers as it may be counterproductive.
> Main objective of FireMonkey has always been security & performance.

I think the discussion is still in this track. I'm trying to be convinced about the performance advantage of FM (mainly because providing higher compatibility is being rejected based on that) and also security (but I don't really care so much about this point because I don't see anything relevant that can be exposed). But of course you don't NEED to answer me if you don't want to.

**Comment by erosman** [comment]
> Btw, I think VM table in FM Help is wrong. For instance, apparently it supports @updateURL and @author. The default metadata block when you click to create a new userScript even uses @author.

Data was based on [Metadata Block - Violentmonkey](https://violentmonkey.github.io/api/metadata-block/). :man_shrugging: 

> Right, but I still don't know how a page could access GM* functions if they aren't explicitly exposed by the UserScript (like by unsafeWindow.GM_getValue = GM_getValue as I said).

The userscript in `page` context can not directly communicate with GM functions on the background script of the extension. There is no direct path by intentional design. Modern browser architecture was especially changed to create these isolation between page & content & browser with no path to exploit, for security reasons.

In order to implement GM functions from `page` context, a `window` messaging & listener must be used to communicate between `page` JavaScript and add-on background script.

For example a  [Window.postMessage()](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) or  [window.dispatchEvent()](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent) etc to be listened to and then applied as GM functions. 
More info: FireMonkey Help under **Receiving data from page scripts**

Above `window` event can also be sent (or listened to, or hijacked) by website's `page` JavaScript to initiate GM functions of the extension.

That was a very simplified short example and explanation.

TM correctly does not provide any GM functions to userscripts injected into `page` context (in other words, only injects into `page` context when `grant` is `none`).

> I think the discussion is still on this track.

Pointing out issues of other similar products, is best left to independent examiners.

**Comment by pintassilgo** [comment]
At least on VM, if an UserScript send GM* functions to the page, the page can execute them straightforward. For example, an script with
```
...
// @grant GM_setValue
// @inject-into page
...

unsafeWindow.GM_setValue = GM_setValue;
```
the page can simply use `GM_setValue('foo', 'bar')` and the value will be stored. The `inject-into` isn't even needed with default settings (`auto` injection mode) if the page doesn't use CSP. Only UJS loaded in `content` context (and `userScripts` of course) prevents any possibility to expose GM* functions to the page.

As in this example, for the page to access GM* functions you need to have this type of vulnerable UserScript installed. GM* are part of the extension and the user is responsible for the UserScripts he installs. Afaik, it's possible for an UserScript to be injected into page context + to use GM* functions + to not expose them to the page, either directly or by messaging/events. So I really don't see VM approach as having security hole.

> Pointing out issues of other similar products, is best left to independent reviewers.

I'm just trying to understand your claims.
- You said "performance difference is immense" between FM and other UJS managers because of userScripts API, but I don't know how could I check this and I fear there's no difference in real world.
- You said that "*VM injects into page WITH GM functions which is a considerable security risk as web sites will also have access to those GM functions*". But I still think that unless the script deliberately exposes GM* functions to the page (either directly or by messaging/listeners), it can safely use GM* functions without the risk of page having access to them. So the risk of VM is if the user install malicious or bad written UserJS, but at the end every manager has almost this same risk.
- And I also don't how, let's say, a VM's UserScript could exploit the difference between `content` and `userScript` contexts.

Sorry if I'm looking pushy, I'm not fluent in English so I may not be choosing the best words and I'm also not sure if I'm being clear, that's why I'm insisting, because I haven't seen answers for some things I asked. But thank you for your work and your time to try to explain things on all these replies.

**Comment by rubaboo** [comment]
> we're basically restricting FireMonkey to JS-skilled users, besides the hassle of having to fix scripts every time they update.

I'm ok with that.

> I can fix scripts that are incompatible with FM. But I'm thinking on those who can't. And even those who can fix would need to reapply the patch every time the script updates.

Not sure if you are thinking about "those who can't" or about your convenience. Those who can't, as well as those who don't want to bother,  should give up playing with a Userscript manager.

> the issue is in malicious/bad written UserScript
> for the page to access GM* functions you need to have this type of vulnerable UserScript installed. GM* are part of the extension and the user is responsible for the UserScripts he installs.

This is the best argument for why `inject-into` should be part of the script and not a "set and forget" attribute.

@pintassilgo,  please open a new issue if you still have one.

**Comment by pintassilgo** [comment]
> Those who can't, as well as those who don't want to bother, should give up playing with a Userscript manager.

I strongly disagree with that. In general, UserScripts have always been considered as simpler and *cross-browser* extensions. It's not needed to know JS to use extensions and the same should apply to UserScripts. I didn't understand JS in my first years as Greasemonkey user (2005 or 2006).

In my opinion, FM should try to be compatible with UserScripts written for GM3 if it doesn't hurt FM performance. No one would deny that GM3 is the closest to *the standard* for UserScripts, right? Even if GM4 doesn't support GM_* functions.

> This is the best argument for why inject-into should be part of the script and not a "set and forget" attribute.

I think the opposite. Any UserScript author could publish an update changing unnecessarily to `@inject-into page` and adding malicious/bad code. If you can define `@inject-into` globally or per-script with higher priority than UserScript metadata, then you may have an extra layer of protection. But the truth is that I'm not really concerned about security as I see no real risk. My point is to enhance compatibility.

**Comment by erosman** [comment]
> At least on VM, if an UserScript send GM* functions to the page, the page can execute them straightforward.

It is not safe.

>  it's possible for an UserScript to be injected into page context + to use GM* functions + to not expose them to the page, either directly or by messaging/events. 

It is not safe.
VM is the only user-script manger doing this.
TM|GM3|GM4|FM are not doing it, therefore the practice is not compatible with other user-script managers.

> [Tampermonkey • Documentation](https://www.tampermonkey.net/documentation.php#_grant)
> 
> If `@grant` is followed by 'none' the sandbox is disabled and the script will run directly at the page context. In this mode no GM_* function but the GM_info property will be available.

> You said "performance difference is immense" between FM  ...

As I said, it is not appropriate to comment on specific issues of competing products as it can be perceived as being biased. Users are free to seek independent review and choose the preferred product.

> But I still think that unless the script deliberately exposes ...

It is not appropriate to post exploits in open discussion.

> And I also don't how, let's say, a VM's UserScript could exploit the difference between content and userScript contexts.

Why do you think Mozilla engineers created `userScripts` API?
Don't you think there was a reason to create an especially made API just for userscripts?
Why did they bother?
Why didn't they leave it to `contentScripts` API?

They created `userScripts` API since there is an inherent security risk with injecting user-scripts into sandboxed `content` context meant for extension's OWN (not 3rd party user-script) code. I was actually the one who brought up the issue to Mozilla and present when discussing the security issues regarding user-scripts, and further liaised with Mozilla engineers during the development and testing of the API. I actually developed FireMonkey (initially for my own use) because of the security issues and AFAIK, FM is the only add-on that uses the secure userScripts API.

Furthermore, `page` context is not even as secure as `content` context and thus has far increased security risks.
As mentioned above, it is not appropriate to post exploits in open discussion.

> In my opinion, FM should try to be compatible with UserScripts written for GM3

AFAIK, it should be compatible with GM3 userscripts. GM3 also didn't provide GM functions to `page` context.
There is a partial compatibility AFA `GM_getValue` & `GM_listValues` but that is due to the architecture of modern browsers.


### Footnotes

There are a number of user-script mangers available with varying features and specifics. Users will choose the one that suits them most. Creating an identical clone of another extension is illogical.

A quote from VM developer 

> `https://github.com/violentmonkey/violentmonkey/issues/869#issuecomment-584746320`
> Just because TM has something doesn't mean we should copy it.

### Finally

The focus of this issue (_Script doesn't work in Firemonkey_) has been resolved. Supplementary issues would be better discussed in a separate issue.

@rubaboo I have made some improvements to the `@inject-into` process which will be in the next v2.14. It should not cause any issues but please let me know once v2.14 is out.

**Comment by erosman** [comment]
> it would be great if the sidebar was pinnable. In a desktop browser it makes sense to have it visible at all times, as before.

Added option to pin drop-down menu for v2.14

**Comment by erosman** [comment]
### Performance Test

**UserScript**
- Load the same userscript/s in FM|GM|TM|VM
- Open `about:performance`
- Go to a page that above runs on
- Compare

**UserStyle/UserCSS**
- Load the same userStyle/userCSS in FM|Stylish|Stylus|xStyle
- Open `about:performance`
- Go to a page that above runs on
- Compare
